Add perf option to benchmark runner script, create text profiles and flamegraphs

apolcyn opened the pull request:
adds a "perf" arg to the `tools/run_tests/run_performance_tests.py` script. 

With this right now, with the "perf" arg, the benchmark script runs the qps worker under perf, with the perf args passed in. It then creates perf text reports and flamegraphs. 

E.g., configured the perf adhoc job to use an arg "--perf=record -F 99 -g".
See reports created https://grpc-testing.appspot.com/job/gRPC_performance_adhoc/HTML_Report/
, from run at https://grpc-testing.appspot.com/job/gRPC_performance_adhoc/185/console

Note currently, flamegraph generation won't work if the arg doesn't include "-g", but text reports will. (thinking we might also want to run "--perf=stat ...")

made temporary changes to master workers for the time being, but after this, need to follow up by adding to benchmark driver machine creation procedure to have flame graph deps installed.

Small note: the arg needs to be wrapped, e.g.: `run_performance_tests.py ... "--perf=record -F 99 -g" ...`

apolcyn commented:
FTR it's probably best if this waits to sample/count only during the benchmark period, as is this samples the whole lifetime of a qps worker

apolcyn commented:
heads up there are some changes that need to be made for multi-scenario

ctiller commented:
Can you test this with some C++ scenarios and post?
(want to make sure that we've got it working in those two cases).

apolcyn commented:
a c++ run with a couple of protobuf qps scenarios is waiting on perf adhoc (#190) https://grpc-testing.appspot.com/job/gRPC_performance_adhoc/

apolcyn commented:
btw there are some go profiles of unary and streaming qps on https://grpc-testing.appspot.com/job/gRPC_performance_adhoc/HTML_Report/
, waiting on c++

apolcyn commented:
C++ unary and streaming benchmark profiles are in https://grpc-testing.appspot.com/job/gRPC_performance_adhoc/HTML_Report/ btw, from the latest run #191 on perf adhoc

jtattermusch commented on tools/run_tests/jobset.py:
```diff
@@ -232,6 +232,7 @@ def start(self):
     cmdline = self._spec.cmdline
     if measure_cpu_costs:
       cmdline = ['time', '--portability'] + cmdline
+    print(' cmdline is ' + repr(cmdline))
```
remove this.

jtattermusch commented on tools/run_tests/performance/process_local_perf_reports.sh:
```diff
@@ -0,0 +1,13 @@
+echo "running perf script on $PERF_DATA_FILE"
+perf script -i $PERF_DATA_FILE > $PERF_SCRIPT_OUTPUT
+
+~/FlameGraph/stackcollapse-perf.pl --kernel $PERF_SCRIPT_OUTPUT | ~/FlameGraph/flamegraph.pl --color=java --hash > ${OUTPUT_DIR}/${OUTPUT_FILENAME}.svg
```
where does the ~/FlameGraph/  come from?   I guess cloning it should be added to the init script?

jtattermusch commented on tools/run_tests/performance/process_local_perf_reports.sh:
```diff
@@ -0,0 +1,13 @@
+mkdir -p $OUTPUT_DIR
```
copyright header missing

jtattermusch commented on tools/run_tests/performance/process_remote_perf_reports.sh:
```diff
@@ -0,0 +1,30 @@
+mkdir -p $OUTPUT_DIR
```
copyright header missing

jtattermusch commented on tools/run_tests/performance/process_remote_perf_reports.sh:
```diff
@@ -0,0 +1,30 @@
+echo "creating perf text report on remote host $USER_AT_HOST"
+
+# "perf report -i" to not wait for terminal input
+ssh $USER_AT_HOST "cd ~/performance_workspace/grpc && perf report -i $PERF_DATA_FILE -v --header > $OUTPUT_FILENAME"
```
pipe the output to gzip so you don't need a separate gzip command?

jtattermusch commented on tools/run_tests/performance/remote_host_build.sh:
```diff
@@ -33,4 +33,5 @@ set -ex
 cd $(dirname $0)/../../..
 
 # execute the build script remotely
+
```
revert changes to this file

jtattermusch commented on tools/run_tests/run_performance_tests.py:
```diff
@@ -80,24 +85,34 @@ def kill(self):
+    ssh_cmd.extend([str(user_at_host), 'cd ~/performance_workspace/grpc/ && %s' % ' '.join(cmdline)])
+    cmdline = ssh_cmd
+
+  print("cmdline for qpsworker is: " + repr(cmdline))
```
remove this line?

jtattermusch commented on tools/run_tests/run_performance_tests.py:
```diff
@@ -80,24 +85,34 @@ def kill(self):
+  if perf_cmd:
+    perf_data_base_name = '%s-%s' % (host_and_port, shortname)
+    # specify -o output file so perf.data gets collected when worker stopped
+    cmdline = perf_cmd + ('-o %s-perf.data' % perf_data_base_name).split(' ') + cmdline
```
nit:   ['-o', '%s-perf.data' % perf_data_base_name] instead of splitting?

jtattermusch commented on tools/run_tests/run_performance_tests.py:
```diff
@@ -57,15 +59,18 @@
   """Encapsulates a qps worker server job."""
 
-  def __init__(self, spec, language, host_and_port):
+  def __init__(self, spec, language, host_and_port, perf_data_base_name=None):
```
nit: perf_db_name is enough? clearly a question of taste, so feel free to ignore.

apolcyn replied:
realized this might be confusing, with it being unrelated to a data base, but just referring to the base name of the "perf.data" file. Renamed to `perf_file_base_name`, open to change though.

jtattermusch commented on tools/run_tests/run_performance_tests.py:
```diff
@@ -275,11 +290,32 @@ def create_qpsworkers(languages, worker_hosts):
+          % (output_filename, _PERF_REPORT_OUTPUT_DIR, perf_base_name)
+
+  return jobset.JobSpec(cmdline=cmd,
+                        timeout_seconds=3 * 60,
```
3*10  (no spaces) when using a named arg

jtattermusch commented on tools/run_tests/run_performance_tests.py:
```diff
@@ -471,11 +543,26 @@ def finish_qps_workers(jobs):
+  report_utils.render_perf_profiling_results('%s/index.html' % _PERF_REPORT_OUTPUT_DIR, profile_output_files)
 
+if total_scenario_failures > 0 or qps_workers_killed > 0:
+  print ("%s scenarios failed and %s qps worker jobs killed" % (total_scenario_failures, qps_workers_killed))
```
nit: no space after print() and change to single quotes

jtattermusch commented on tools/run_tests/run_performance_tests.py:
```diff
@@ -471,11 +543,26 @@ def finish_qps_workers(jobs):
-if total_scenario_failures > 0 or qps_workers_killed > 0:
-  print ("%s scenarios failed and %s qps worker jobs killed" % (total_scenario_failures, qps_workers_killed))
+if perf_report_failures > 0:
+  print ("%s perf profile collection jobs failed" % perf_report_failures)
```
same as above

jtattermusch commented on tools/run_tests/run_performance_tests.py:
```diff
@@ -49,6 +49,8 @@
 import traceback
 import uuid
+import report_utils
+from distutils.spawn import find_executable
```
you don't need find_executable. please remove.

jtattermusch commented on tools/run_tests/run_performance_tests.py:
```diff
@@ -403,6 +464,8 @@ def finish_qps_workers(jobs):
                   action='store_const',
                   const=True,
                   help='Run netperf benchmark as one of the scenarios.')
+argp.add_argument('--perf',
```
maybe --perf_args would be more self-explanatory here?

apolcyn commented:
job #192 on perf adhoc is waiting on this btw. I can change the perf_adhoc config back to normal if needed though. Right now, there is a "--perf_args" option passed to it.

jtattermusch commented:
Please let me know once this is ready for another review pass.

apolcyn commented:
I tried to address all of the comments in the latest commits. I think this should be ready for another pass now

apolcyn commented:
btw another run is on perf adhoc (#193), the perg_args is now "--perf_args=record -F 997 -g", last time it was at -F 997, which I think is why the output was sparse

jtattermusch commented on tools/run_tests/run_performance_tests.py:
```diff
@@ -471,11 +540,26 @@ def finish_qps_workers(jobs):
     finally:
       # Consider qps workers that need to be killed as failures
       qps_workers_killed += finish_qps_workers(scenario.workers)
+      if perf_cmd and scenario_failures == 0:
```
why do this in the finally block? 
1. you are generating the flamegraph only if there were no scenario failures anyway.
2. throwing in the finally block is not a good practice in general (in case there was and exception in the try block it eats it a throws a different one instead of it)

just run your code in the try block or after the entire try-finally.

jtattermusch commented on tools/run_tests/run_performance_tests.py:
```diff
@@ -433,7 +493,13 @@ def finish_qps_workers(jobs):
+if args.perf_args:
+  # Expect /usr/bin/perf to be installed here, as is usual
+  perf_cmd = ['/usr/bin/perf'] 
+  perf_cmd.extend(args.perf_args.split(' '))
```
if I put two spaces in the perf_args, will that result in and empty items in the list after splitting?   
this is a bit errorprone

jtattermusch commented on tools/run_tests/performance/build_performance.sh:
```diff
@@ -40,7 +40,10 @@ CONFIG=${CONFIG:-opt}
-make CONFIG=${CONFIG} EMBED_OPENSSL=true EMBED_ZLIB=true qps_worker qps_json_driver -j8
+if [ "$ON_REMOTE" != "true" ]
+then
+	make CONFIG=${CONFIG} EMBED_OPENSSL=true EMBED_ZLIB=true qps_worker qps_json_driver -j8
```
this skips qps_worker build if running remotely -> will break C++?

jtattermusch commented on tools/run_tests/run_performance_tests.py:
```diff
@@ -462,7 +462,23 @@ def run_collect_perf_profile_jobs(hosts_and_base_names, scenario_name):
 argp.add_argument('--perf_args',
-                  help='Wrap QPS workers in a perf command, with arguments to perf specified here.')
+                  help=""" 
+                          Example usage: "--perf_args=record -F 99 -g".
```
won't the indentation be included in the resulting string? isn't there a better way in python. I'd avoid """ if possible.

apolcyn replied:
yeah it did turn out that indentation was included in the result, updated without using """

jtattermusch commented:
Also, you need to resolve a conflict.

apolcyn commented:
needed to move the perf profile collection to after qps workers killed, just since the perf output gets created only when the job ends. 

even with -F 997, profiles still look a bit sparse, but took out the arguments to the scripts, re-running on perf adhoc as #195

jtattermusch commented on tools/run_tests/run_performance_tests.py:
```diff
@@ -435,7 +510,13 @@ def finish_qps_workers(jobs):
+perf_cmd = None
+if args.perf_args:
+  # Expect /usr/bin/perf to be installed here, as is usual
+  perf_cmd = ['/usr/bin/perf'] 
```
nit: do you really need `/usr/bin/perf` here and not just `perf`?

apolcyn replied:
I had it at `/usr/bin/perf` onyl because the qps worker command is running with shell=False (so it doesn't get the PATH or find `perf` I take it, and what I saw).

Maybe qps worker commands should be changed to run using the shell?

apolcyn commented:
cc @sreecha there are flame graphs from unary and streaming qps benchmarks in https://grpc-testing.appspot.com/job/gRPC_performance_adhoc/HTML_Report/, but they still look a bit sparse to me, do these match at all what you saw?

apolcyn commented:
This has already been approved, but it's changed somewhat significantly since then to fix issues with the sparse flame graphs. Is this ok to merge? PLMK

ctiller commented:
Jenkins: test this please

apolcyn commented:
latest updates rebase and squash.

previously these test faillures were occuring:
- interop PR: https://github.com/grpc/grpc/issues/9011, fixed in https://github.com/grpc/grpc/pull/8845
- linux PR: https://github.com/grpc/grpc/issues/9014
- performance PR: https://github.com/grpc/grpc/issues/8967#event-882988054, fixed in https://github.com/grpc/grpc/pull/8969
- performance PR:
